src/Angular.js

```diff
@@ -1772,6 +1772,48 @@ function getTestability(rootElement) {
   return injector.get('$$testability');
 }
 
+function rootElementProviderFactory(rootElement) {
+  return ['$shutdownProvider', function($shutdownProvider) {
+    $shutdownProvider.register(function() {
+      if (rootElement.dealoc) {
+        rootElement.dealoc();
+      } else {
+        rootElement.find('*').removeData();
+        rootElement.removeData();
```

Comment:
This is not enough as it won't remove event listeners. How about:
```js
jqLite.cleanData(rootElement.find('*'));
jqLite.cleanData(rootElement);
```

or something analogous to @gkalpak's code but using `cleanData` if we want to not break `ng-non-bindable`-contained AngularJS apps.

This, BTW, made me notice that jqLite's implementation of `removeData` is really broken. It should only remove user-set data and not touch the event handlers storage at all. I reported #15869 to fix that in 1.7.